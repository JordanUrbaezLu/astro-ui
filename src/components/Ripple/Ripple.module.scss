.container {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;

  span {
    animation-name: ripple;
    border-radius: 100%;
    opacity: 0.75;
    position: absolute;
    transform: scale(0);
  }

  @keyframes ripple {
    to {
      opacity: 0;
      transform: scale(2);
    }
  }
}
