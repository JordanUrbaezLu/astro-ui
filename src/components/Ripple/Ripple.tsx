import classNames from "classnames";
import * as React from "react";
import styles from "./Ripple.module.scss";

export interface RippleInstanceCoordinates {
  x?: number;
  y?: number;
  size?: number;
}

export type RipplePosition = "center" | "cursor";

export interface RippleProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The color of the Ripple
   */
  color?: string;
  /**
   * The duration of the Ripple
   */
  duration?: number;
  /**
   * The position of the Ripple relative to its container
   *
   * @default "cursor"
   */
  position?: RipplePosition;
}

export const Ripple: React.FC<RippleProps> = (props) => {
  const {
    className,
    color = "#381438",
    duration = 1300,
    position = "cursor",
    ...rest
  } = props;
  const [rippleArray, setRippleArray] = React.useState<
    Array<RippleInstanceCoordinates>
  >([]);

  const addRipple = (event: React.MouseEvent) => {
    const rippleContainer =
      event.currentTarget.getBoundingClientRect();
    const size =
      rippleContainer.width > rippleContainer.height
        ? rippleContainer.width
        : rippleContainer.height;

    let x, y;

    if (position === "cursor") {
      x = event.pageX - rippleContainer.x - size / 2;
      y = event.pageY - rippleContainer.y - size / 2;
    } else {
      x = 0;
      y = 0;
    }

    const newRipple: RippleInstanceCoordinates = {
      x,
      y,
      size,
    };

    setRippleArray((prevRipples) => [...prevRipples, newRipple]);
  };

  return (
    <div
      className={classNames(styles.container, className)}
      onMouseDown={addRipple}
      role="presentation"
      {...rest}
    >
      {rippleArray.length > 0 &&
        rippleArray.map(
          (ripple: RippleInstanceCoordinates, index: number) => {
            return (
              <span
                key={"span" + index}
                role="presentation"
                style={{
                  animationDuration: `${duration}ms`,
                  backgroundColor: color,
                  top: ripple.y,
                  left: ripple.x,
                  width: ripple.size,
                  height: ripple.size,
                }}
              />
            );
          }
        )}
    </div>
  );
};
