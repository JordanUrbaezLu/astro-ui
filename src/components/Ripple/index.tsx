/* istanbul ignore file */
export * from "./Ripple";
