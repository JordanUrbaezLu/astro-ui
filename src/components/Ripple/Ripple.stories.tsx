import { Ripple } from "./Ripple";

export default {
  component: Ripple,
  title: "Components/Utilities/Ripple",
};

export const Default = () => <Ripple />;
