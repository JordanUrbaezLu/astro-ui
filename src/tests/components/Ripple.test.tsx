import { render, fireEvent, screen } from "@testing-library/react";
import { Ripple } from "../../components/Ripple/Ripple";
import { axe } from "jest-axe";

describe("Ripple", () => {
  test("Should render component corectly.", () => {
    const { container } = render(<Ripple />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(<Ripple />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should addRipple on userEvent.", () => {
    render(<Ripple duration={10000} />);

    fireEvent.mouseDown(screen.getByRole("presentation"));

    expect(
      screen.getAllByRole("presentation")[1]
    ).toBeInTheDocument();
  });
});
