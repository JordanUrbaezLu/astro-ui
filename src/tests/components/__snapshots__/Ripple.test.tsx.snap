// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple Should render component corectly. 1`] = `
<div
  class="container"
  role="presentation"
/>
`;
